tests/virt_autotest/libvirt_virtual_network_init.pm

```diff
@@ -53,23 +53,22 @@ sub run_test {
     #Install required packages
     zypper_call '-t in iproute2 iptables iputils bind-utils sshpass nmap';
 
-    #Check with Guest status before libvirt virtual network tests
-    virt_autotest::virtual_network_utils::check_guest_status();
-
     #Prepare Guests
     foreach my $guest (keys %xen::guests) {
         #Archive deployed Guests
         #NOTE: Keep Archive deployed Guests for restore_guests func
         assert_script_run("virsh dumpxml $guest > /tmp/$guest.xml");
         upload_logs "/tmp/$guest.xml";
-        #Start installed Guests
-        assert_script_run("virsh start $guest", 60);
-        #Wait for forceful boot up guests
-        sleep 60;
+        #Used with attach-detach(hotplugging) interface to confirm all virtual network mode
+        #NOTE: Required all guests keep running status
+        #Check that all guests are still running before virtual network tests
+        script_retry("nmap $guest -PN -p ssh | grep open", delay => 30, retry => 6, timeout => 180);
         save_guest_ip($guest, name => "br123");
-        my $mode            = is_sle('=11-sp4') ? '' : '-f';
-        my $default_ssh_key = "/root/.ssh/id_rsa.pub";
-        exec_and_insert_password("ssh-copy-id -i $default_ssh_key -o StrictHostKeyChecking=no $mode root\@$guest");
+        if (!(get_var('VIRT_AUTOTEST'))) {
+            my $mode            = is_sle('=11-sp4') ? '' : '-f';
+            my $default_ssh_key = "/root/.ssh/id_rsa.pub";
+            exec_and_insert_password("ssh-copy-id -i $default_ssh_key -o StrictHostKeyChecking=no $mode root\@$guest");
+        }
```

Comment:
```suggestion
        my $mode            = is_sle('=11-sp4') ? '' : '-f';
        my $default_ssh_key = (!(get_var('VIRT_AUTOTEST'))) ? "/root/.ssh/id_rsa.pub" : "/var/testvirt.net/.ssh/id_rsa.pub";
        exec_and_insert_password("ssh-copy-id -i $default_ssh_key -o StrictHostKeyChecking=no $mode root\@$guest");
```

Reply:
Ideally, can you make `virt_autotest::virtual_network_utils::ssh_copy_id();` function out of this code please?

Reply:
as my mentioned, setup_dns_service, had created SSH KEY and copy it to guest system. So, I do not plan to copy it to guest system again as duplicated for QA part. So, the current code SKIP copy SSH key to guest for QA part.

Reply:
> As my mentioned, setup_dns_service, had created SSH KEY and copy it to guest system. So, I do not plan to copy it to guest system again as duplicated for QA part.

As I mentioned before, we should not create duplicates for the same. That's why we should create `virt_autotest::virtual_network_utils::ssh_copy_id();` function from the code above and use it here as well as in `setup_dns_service`.

Reply:
Great, create  ssh_copy_id func as below: 
- used with `virt_autotest::virtual_network_utils::ssh_copy_id();`
JUST QAM copy the SSK Key to guest system
`
sub ssh_copy_id {
    if (check_var('TEST', 'qam-xen-networking')) {
    my $mode            = is_sle('=11-sp4') ? '' : '-f';
    my $default_ssh_key = "/root/.ssh/id_rsa.pub";
    exec_and_insert_password("ssh-copy-id -i $default_ssh_key -o StrictHostKeyChecking=no $mode root\@$guest");
    }
}
`
Do you agree?

Reply:
No. As you said you also copy the SSH key to the system.

> as my mentioned, setup_dns_service, had created SSH KEY and copy it to guest system. 

I suggest this:

```perl
sub ssh_copy_id {
    my $mode = is_sle('=11-sp4') ? '' : '-f';
    my $default_ssh_key = (!(get_var('VIRT_AUTOTEST'))) ? "/root/.ssh/id_rsa.pub" : "/var/testvirt.net/.ssh/id_rsa.pub";
    if (script_run("ssh -o PreferredAuthentications=publickey root\@$guest hostname -f") != 0) {
        exec_and_insert_password("ssh-copy-id -i $default_ssh_key -o StrictHostKeyChecking=no $mode root\@$guest");
    }
```

Reply:
Awesome, make sense here about SSH Keys setting both us. I am working on doing validation run with your suggestion now.

Reply:
@pdostal Finished the latest validation run with the current code changed:
 gi-guest_developing-on-host_developing-kvm
http://10.67.131.5/tests/192

Reply:
@Julie-CAO Can you help to review whether the above agreed solution for keys work for our tests?

Reply:
If I understand the topic correctly, Pavel hope to extract the copy_ssh_id() function from the pm and move it to the lib to be shared by both QAM and setup_dns from QA. I just went through setup_dns which is created by Wayne and found it achieves the passwordless access by other means --- modifying the ssh_config file. As the copy_ssh_id() function is expected to be used by QAM only, I think the solution(the code and the location) here is ok.

Reply:
Actually, setup_dns calls a shell script to copy the ssh key to guest rather than Perl script. My previous description may not be precise, but the conclusion should be correct  ---- copy_ssh_id() is not expected to be used by setup_dns.

Reply:
@Julie-CAO @alice-suse thank you for the analysis. This solution should work in all our cases and should not interfere with your shell script. My goal is to move those procedures we all do in multiple scenarios into function so we all can use them instead duplicating the code in hidden shell scripts.

Reply:
Fully understand and agree!